### PR TITLE
Resolve wrongly merging bond/angle/dihedral types

### LINKED ISSUE #4001: LAMMPS bond/angle/dihedral types are wrongly deleted
## Expected behavior ##

Correct number of bond/angle/dihedral types are written to lammps data file.

## Actual behavior ##

Method, core.topologyobjects._removeDupes(), wrongly removes bond/angle/dihedral types. This function is written for cases where the bond/angle/dihedral types are defined as a tuple of atom types, so that if both bond (1,2) and (2,1) are present, the two will be merged. However, in lammps the types are defined as integers in string format, so if there are more than 20 types, then bond 12 and 21 will be merged.

### PR DESCRIPTION
**Pull request by jaclark5**
Fixes #4001

Changes made in this Pull Request:
 - if statement before calling _removeDupes() to make sure the bond/angle/dihedral type names are tuples, otherwise this step is skipped.

PR Checklist
------------
 - [X] Tests?
 - [NA] Docs?
 - [X] CHANGELOG updated?
 - [X] Issue raised/referenced?

### CODE REVIEW

**Review by hmacdope — CHANGES_REQUESTED** [review]
@jaclark5 sorry I have been so slow to review. Would you be able to add a test to check the new behaviour is correct?

**Comment by jaclark5** [comment]
I could use some help understanding why the test I wrote isn't working. Basically I import a LAMMPS data file with 22 bond types. 

1. The first assertion confirms that there are indeed 22 bond types since the bond types aren't defined with tuples an _removeDupes is not called. 
2. I then call `universe.atoms.bonds.topDict._removeDupes()` which will combine bond type 12 and bond type 21 to return 21 bond types. I then assert that the number of bonds does not equal 22.

However, it seems that the `TopologyDict` attribute `dict` doesn't actually update to reflect the changes made in step 2.
I added some print statements to the TopologyDict to try to debug this, here is the output from `pytest test_topologyobjects.py::TestTopologyGroup::test_bond_no_reversal -vvv`
```
Creating a TopologyDict Instance
Initiating TopologyDict
topDict.keys:         topDict id: 4589672672, topDict.dict id: 4590454144, len(topDict.dict.keys()): 22
Creating a TopologyDict Instance
Initiating TopologyDict
topDict._removeDupes: topDict id: 4589672672, topDict.dict id: 4589549056, len(topDict.dict.keys()): 21
Creating a TopologyDict Instance
Initiating TopologyDict
topDict.keys:         topDict id: 4589672672, topDict.dict id: 4590379904, len(topDict.dict.keys()): 22
```

Edit: I figured it out, it appears that the class is reinitiated every time the bond types are called. Probably because `TopologyGroup.topDict` returns `TopologyDict(self)` instead of checking if it exists. This is a separate issue, should I raise it?

**Comment by hmacdope** [comment]
I think `bondDict` and its cacheing is due to be removed with #3777, but I agree this is perhaps a bit of a "dark-arts" part of MDAnalysis. Well done for figuring out your issue.

There is also the related issue of the current system of bonds and even the solution in #3777 only retaining one bond with identical  indices (eg only one of 1-2 or 2-1 can exist). This is documented in #3818  and I have been meaning to get on with it as it is definately an issue. I think the semi-conclusion  we came to last time was that this is a topology parser issue, IE the parser should sort out the bonds so that there are no duplicates. 

Sorry for the info dump but thought I would help out with some context.

**Comment by jaclark5** [comment]
I'm not sure it will since that PR doesn't touch topologyobjects.py. The issue is that `TopologyGroup.topDict` is supposed call `TopologyDict` and then cache it so that it doesn't keep reinitiating a new instance. This could be resolved by setting something like `self._topDict = None` in `TopologyGroup.__init__` and then changing `TopologyGroup.topDict` from:
```
@property
@cached('dict')
def topDict(self):
    return TopologyDict(self)
```
to
```
@property
def topDict(self):
    if self._topDict is None:
        self._topDict = TopologyDict(self)
    return self._topDict
```

or just initiate topDict in `TopologyGroup.__init__`... but I haven't looked through to understand the order of operations that would justify the current pattern with a cached method acting as an attribute instead of defining an attribute to begin with.

File: package/MDAnalysis/core/topologyobjects.py [hunk 1]
```diff
@@ -462,6 +464,7 @@ def _removeDupes(self):
                 newdict[k[::-1]] += self.dict[k]
 
         self.dict = newdict
+        print("topDict._removeDupes: topDict id: {}, topDict.dict id: {}, len(topDict.dict.keys()): {}".format(id(self), id(self.dict), len(self.dict)))
```
Comment (hmacdope) [inline package/MDAnalysis/core/topologyobjects.py]: remove

File: package/MDAnalysis/core/topologyobjects.py [hunk 2]
```diff
@@ -473,6 +476,7 @@ def __len__(self):
 
     def keys(self):
         """Returns a list of the different types of available bonds"""
+        print("topDict.keys:         topDict id: {}, topDict.dict id: {}, len(topDict.dict.keys()): {}".format(id(self), id(self.dict), len(self.dict)))
```
Comment (hmacdope) [inline package/MDAnalysis/core/topologyobjects.py]: Remove

File: package/MDAnalysis/core/topologyobjects.py [hunk 3]
```diff
@@ -656,6 +660,7 @@ def topDict(self):
 
         .. versionadded 0.9.0
         """
+        print("Creating a TopologyDict Instance")
```
Comment (hmacdope) [inline package/MDAnalysis/core/topologyobjects.py]: Remove

**Review by hmacdope — CHANGES_REQUESTED** [review]
I think this seems reasonable approach for now, I guess you found that the original setup is error prone and likely still is, but this is a step in the right direction

**Comment by jaclark5** [comment]
Hey @hmacdope, I was wrong, the issue isn't that the `TopologyDict` isn't caching correctly in the `TopologyGroup`, but instead that `topologyattrs._Connection.get_atoms` reinitializes the `TopologyGroup` every time you call the `AtomGroup.bonds` and this is not resolved in #3777 but it could be. Once that is resolved and pushed my tests will pass.

**Comment by hmacdope** [comment]
@jaclark5 do we still need #3777 or is this now ready for a seperate review? I think the approach taken here can stand on its own. :)

File: package/MDAnalysis/core/topologyobjects.py [hunk 4]
```diff
@@ -445,7 +445,8 @@ def __init__(self, topologygroup):
             except KeyError:
                 self.dict[btype] = [b]
 
-        self._removeDupes()
+        if self.dict and isinstance(list(self.dict.keys())[0], tuple):
```
Comment (hmacdope) [inline package/MDAnalysis/core/topologyobjects.py:455]: Comment here as to why the `if` is needed would be good.

File: testsuite/MDAnalysisTests/core/test_topologyobjects.py [hunk 5]
```diff
@@ -300,6 +301,12 @@ def test_bond_reversal(self, PSFDCD, b_td):
         tg2 = b_td[b[::-1]]
         assert tg1 == tg2
 
+    def test_bond_no_reversal(self):
```
Comment (hmacdope) [inline testsuite/MDAnalysisTests/core/test_topologyobjects.py:307]: Again a quick comment on why this tests what it tests, eg `21` vs `12`.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Two small comments but other than that I am happy.

**Review by hmacdope — APPROVED** [review]
Thanks @jaclark5 for the fantastic work!

**Comment by hmacdope** [comment]
@MDAnalysis/coredevs any chance anyone can take a quick peek here also?

**Comment by jaclark5** [comment]
> @jaclark5 do we still need #3777 or is this now ready for a seperate review? I think the approach taken here can stand on its own. :)

@hmacdope Not anymore. I had another test that I removed that also removed that dependency. This other test would call `universe.atoms.bonds.topdict._removeDupes()` and then again call `universe.atoms.bonds.topdict.types()` to show that bond types 12 and 21 were combined. However it failed since calling `universe.atoms.bonds.topdict._removeDupes()` reinitialized the `topdict` and wouldn't retain the effect of calling `universe.atoms.bonds.topdict._removeDupes()`. 

I think this is still something to think about, since it would slow MDAnalysis down to reinitialize the bonds topdict every time it's called. Should I make it another issue or would you want to incorporate that into #3777 ?

**Comment by hmacdope** [comment]
> > @jaclark5 do we still need #3777 or is this now ready for a seperate review? I think the approach taken here can stand on its own. :)
> 
> @hmacdope Not anymore. I had another test that I removed that also removed that dependency. This other test would call `universe.atoms.bonds.topdict._removeDupes()` and then again call `universe.atoms.bonds.topdict.types()` to show that bond types 12 and 21 were combined. However it failed since calling `universe.atoms.bonds.topdict._removeDupes()` reinitialized the `topdict` and wouldn't retain the effect of calling `universe.atoms.bonds.topdict._removeDupes()`.
> 
> I think this is still something to think about, since it would slow MDAnalysis down to reinitialize the bonds topdict every time it's called. Should I make it another issue or would you want to incorporate that into #3777 ?

@jaclark5 raise an issue, I think we can just move forward with this as it is :)

**Comment by jaclark5** [comment]
@MDAnalysis/coredevs I would appreciate another review :)

**Comment by hmacdope** [comment]
@MDAnalysis/coredevs can I get another set of eyes here? If no action in a week I will just go ahead. Sorry we have been a bit slow @jaclark5.

**Comment by hmacdope** [comment]
Cycling CI for aarch.

File: package/CHANGELOG [hunk 6]
```diff
@@ -35,6 +35,7 @@ Deprecations
  * 2.5.0
 
 Fixes
+  * Fix deletion of bond/angle/dihedral types (PR #4003, Issue #4001)
```
Comment (hmacdope) [inline package/CHANGELOG]: @jaclark5 can you move this to 2.6.0?

**Review by hmacdope — CHANGES_REQUESTED** [review]
@jaclark5 just need to move some CHANGELOG entry to 2.6.0

**Comment by hmacdope** [comment]
@IAlibay can I safely ignore CI failure here?

**Comment by IAlibay** [comment]
> @IAlibay can I safely ignore CI failure here?

Ahh sorry for the delay, I thought I had seen a notification come through 😅 

Yeahhh ignore it here, it's not relatetd to this code. It's because of the 1.1 release of contourpy from ~ 3 days ago and the fact that we annoyingly still support 32bit... I'll open a PR.

**Review by hmacdope — APPROVED** [review]
Thanks so much @jaclark5! Great work as always. Great to finally get it in.

**Pull request merged**